### ISSUE TITLE
Parameter annotations in AST are removed by visitor

### ISSUE THREAD

**Issue by shiftinv** [snok/flake8-type-checking#95]
<sup>I read that you're currently on holiday, I mainly just wanted to create these issues for tracking so I don't forget about them - please do enjoy your holiday before anything else, these issues are probably as unimportant as it gets ☺️ </sup>

While debugging the issue that led to #93, I noticed that `ImportVisitor.register_function_annotations` would remove the `annotation` fields from all function parameters:
https://github.com/snok/flake8-type-checking/blob/03e868ce7da81ece39373e622c1ea421662e9b69/flake8_type_checking/checker.py#L676-L678

To illustrate this, one can set a breakpoint before and after this statement, dumping the AST (`node`) and comparing:
https://github.com/snok/flake8-type-checking/blob/03e868ce7da81ece39373e622c1ea421662e9b69/flake8_type_checking/checker.py#L755
```diff
--- tree_orig   2022-07-08 23:36:55.687419100 +0200
+++ tree_new    2022-07-08 23:37:08.295574500 +0200
@@ -5,9 +5,7 @@
       args=arguments(
         posonlyargs=[],
         args=[
-          arg(
-            arg='x',
-            annotation=Name(id='int', ctx=Load()))],
+          arg(arg='x')],
         kwonlyargs=[],
         kw_defaults=[],
         defaults=[]),
```

From what I understand this is required for some checks, and while I unfortunately don't have a concrete example for it, I believe this could break other plugins running afterwards that rely on these annotations, since flake8 reuses the AST for all plugins (at least the version I'm using, 4.0.1, does).

Looking at other issues, something similar seems to have happened in #90, both possibly as a side effect of #89 (?) which removed the copy operation that allowed the code to freely mutate the AST without affecting other flake8 plugins.

I tried a few things in an attempt to make it work and create a PR with a fix that doesn't involve copying the entire AST, but sadly I couldn't manage it with my (very limited) knowledge of how the `ast` module works 😅

**Comment by sondrelg** [comment]
Hey @shiftinv! 🙂 

Yes you're right, it's a bit of a weird problem to fix. The plugin needs that functionality (the removed annotations), the ast tree shouldn't be mutated to break other plugins, and copying the tree is slow and can raise `RecursionError`s in large trees.

I was hoping that v2.0.2 would fix this, by substituting `ast.NodeTransformer` with `ast.NodeVisitor` (#91). Would you be able to test that version to see if the issue persists?

**Comment by shiftinv** [comment]
Hiya!
Yeah, it does seem like quite a tricky issue. It's definitely understandable that copying the tree isn't really viable, I suppose there's a good reason as to why flake8 shares the ast between plugins.
The tests above were done in v2.0.2, and while the switch to `ast.NodeVisitor` fixed a large portion of the mutation issue, the removal of `annotation` fields unfortunately still persists.

**Comment by sondrelg** [comment]
Ok good to know 👍 I'll take another look at this in a bit then. Thanks again 🙂

**Comment by shiftinv** [comment]
Fixed by #104 (thanks!).
